Drivers/Src/MotorController.c

```diff
@@ -1,124 +1,217 @@
+                &err);
+	assertOSError(0, err);
+
+    BSP_CAN_Init(CAN_3, MotorController_CountIncoming, MotorController_Release);
+
+    uint8_t data[8] = {0};
+    memcpy(
+        data,
+        &busCurrentFractionalSetPoint,
+        sizeof(busCurrentFractionalSetPoint)
+    );
+    OSSemPend(&MotorController_MailSem4,
+            0,
+            OS_OPT_PEND_BLOCKING,
+            &ts,
+            &err);
+	assertOSError(0, err);
+    ErrorStatus initCommand = BSP_CAN_Write(CAN_3, MOTOR_POWER, data, MAX_CAN_LEN);
+    if (initCommand == ERROR) {
+		MotorController_Release();
+        Motor_FaultBitmap = T_INIT_FAIL;
+        assertTritiumError(Motor_FaultBitmap);
+	}
 }
 
 /**
- * @brief   Sends MOTOR DRIVE command on CAN2
+ * @brief   Sends MOTOR DRIVE command on CAN3
  * @param   newVelocity desired motor velocity setpoint in m/s
- * @param   motorCurrent desired motor current setpoint as a percentage of max current setting
+ * @param   motorCurrent desired motor current setpoint as a percentage of max current setting (0.0-1.0)
  * @return  None
  */ 
 void MotorController_Drive(float newVelocity, float motorCurrent){
+    CPU_TS ts;
+	OS_ERR err;
 
-    uint32_t nv = *((uint32_t *)((void *) &newVelocity));
-    uint32_t mc = *((uint32_t *)((void *) &motorCurrent));
-
-    uint8_t data[8] = {0};
-    int index = 0;
-    while(index < MAX_CAN_LEN/2){
-        data[index] = (mc >> (8 * (MAX_CAN_LEN/2-index-1))) & 0xFF; //split inputs into bytes
-        index++;
-        
-    }
-    int i = 0;
-    while(index < MAX_CAN_LEN){
-        data[index] = (nv >> (8 * (MAX_CAN_LEN/2-i-1))) & 0xFF;
-        index++;
-        i++;
-    }
+    uint8_t data[8];
+    memcpy(data, &motorCurrent, sizeof(motorCurrent));
+    memcpy(data + sizeof(motorCurrent), &newVelocity, sizeof(newVelocity));
     
-    BSP_CAN_Write(CAN_3, MOTOR_DRIVE, data, MAX_CAN_LEN);
+    OSSemPend(&MotorController_MailSem4,
+			  0,
+			  OS_OPT_PEND_BLOCKING,
+			  &ts,
+			  &err);
+	assertOSError(0, err);
+    ErrorStatus result = BSP_CAN_Write(CAN_3, MOTOR_DRIVE, data, MAX_CAN_LEN);
+    if (result == ERROR) {
+		MotorController_Release();
+	}
+
 }
 
 /**
- * @brief   Reads most recent command from CAN2 bus
+ * @brief   Reads most recent command from CAN3 bus
  * @param   message the buffer in which the info for the CAN message will be stored
  * @return  SUCCESS if a message is read
  */ 
-ErrorStatus MotorController_Read(CANbuff *message, car_state_t *car){
+ErrorStatus MotorController_Read(CANbuff *message){
 
     uint32_t id;
     uint8_t data[8] = {0};
-    uint32_t length = BSP_CAN_Read(CAN_3, &id, data);
     uint32_t firstSum = 0;
     uint32_t secondSum = 0;
+    CPU_TS ts;
+	OS_ERR err;
+	
+	// Check to see if a mailbox is available: BLOCKING
+	OSSemPend(
+        &MotorController_ReceiveSem4,
+	    0,    
+		OS_OPT_PEND_BLOCKING,
+		&ts,
+		&err);
+	assertOSError(0, err);
+    ErrorStatus status = BSP_CAN_Read(CAN_3, &id, data);
```

Comment:
Is it possible for this function to error out and not release the mailbox semaphore?